router/main_test.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+	"time"
+)
+
+func Test_makeHandler(t *testing.T) {
+	config := []struct {
+		RequestHost        string
+		UpstreamURL        string
+		ExpectedStatusCode int
+	}{
+		{
+			RequestHost:        "martindekov.example.xyz/",
+			UpstreamURL:        "http://localhost:8080/myfunction",
+			ExpectedStatusCode: http.StatusInternalServerError,
+		},
+		{
+			RequestHost:        "martindekov.example.xyz/",
+			UpstreamURL:        "/myfunction",
+			ExpectedStatusCode: http.StatusServiceUnavailable,
+		},
+		{
+			RequestHost:        "",
+			UpstreamURL:        "http://localhost:8080/myfunction",
+			ExpectedStatusCode: http.StatusBadRequest,
+		},
+		{
+			RequestHost:        "",
+			UpstreamURL:        "http://localhost:8080/",
+			ExpectedStatusCode: http.StatusBadRequest,
+		},
+	}
+
+	timeout := time.Second * 30
+
+	for _, test := range config {
+		req, err := http.NewRequest(http.MethodPost, "/", nil)
+		if err != nil {
+			t.Fatalf("Request error: `%v`\n", err)
+		}
+
+		URI := strings.TrimPrefix(test.UpstreamURL, "http://localhost:8080")
+		req.RequestURI = URI
+		req.Host = test.RequestHost
+
+		rec := httptest.NewRecorder()
+		c := &http.Client{
+			Timeout: timeout,
+		}
+
+		URL := strings.TrimSuffix(test.UpstreamURL, "myfunction")
+		handlerFunc := makeHandler(c, timeout, URL)
+		handlerFunc(rec, req)
+
+		if rec.Code != test.ExpectedStatusCode {
+			t.Errorf("Expected status code : `%v` got: `%v`", test.ExpectedStatusCode, rec.Code)
+		}
+
+		read, err := ioutil.ReadAll(rec.Body)
```

Comment:
Do you need to call `ioutil.ReadAll` anymore?

Reply:
`rec.Body` is `Body *bytes.Buffer` - correct me if I'm wrong, but I don't think you need to call `ReadAll` on it.

Reply:
That way yeah the response body is `bytes.Buffer` but I need it as `[]byte` to convert it to string later on and compare values. The same approach is in `requesthandler_test.go` in `faas/watchdog`